src/NumericalAlgorithms/RootFinding/GslMultiRoot.cpp

```diff
@@ -0,0 +1,151 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "GslMultiRoot.hpp"
+
+#include <exception>
+#include <gsl/gsl_multiroots.h>
+#include <gsl/gsl_vector.h>
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+
+class rootfinding_exception : public std::exception {
+  virtual const char* what() const throw() { return "GSLMultiRoot failed."; }
+};
+
+void printstate_GSLMultiRootFinder(size_t iter, gsl_multiroot_fdfsolver* s,
+                                   size_t dim, bool PrintHeader = false) {
+  if (PrintHeader){
+    Parallel::printf("Iter\t");
+    for (size_t i = 0; i < dim; ++i) {
+      Parallel::printf(" x[%u]\t", i);
+    }
+    for (size_t i = 0; i < dim; ++i) {
+      Parallel::printf(" f[%u]\t", i);
+    }
+    Parallel::printf("\n");
+  }
+
+  Parallel::printf("%u\t", iter);
+  for (size_t i = 0; i < dim; ++i) {
+    Parallel::printf("%3.4f  ", gsl_vector_get(s->x, i));
+  }
+  for (size_t i = 0; i < dim; ++i) {
+    Parallel::printf("%1.3e  ", gsl_vector_get(s->f, i));
+  }
+  Parallel::printf("\n");
+}
+
+template <size_t Dim>
+std::array<double, Dim> GSLMultiRoot(
+    const GslMultiRootFunctionFdf<Dim>& func,
+    const std::array<double, Dim> initial_guess,
+    const double& absolute_tolerance, const size_t maximum_iterations,
+    const double relative_tolerance, const Verbosity verbosity,
+    const double maximum_absolute_tolerance, const Method method) {
+  gsl_multiroot_function_fdf f = {
+      &Gsl_MultiRootFunctionFdf_Wrapper_f<Dim>,
+      &Gsl_MultiRootFunctionFdf_Wrapper_df<Dim>,
+      &Gsl_MultiRootFunctionFdf_Wrapper_fdf<Dim>, Dim,
+      const_cast<GslMultiRootFunctionFdf<Dim>*>(&func)};
+
+  // Set up initial guess as stored in initial_guess:
+  gsl_vector* gsl_root = gsl_vector_alloc(Dim);
```

Comment:
Make the pointers const where appropriate.

Reply:
I still need to address this